### PR TITLE
Include Multilock for GetAttPreState With Tests

### LINKED ISSUE #7658: Improve getAttPreState with multilock
# 🚀 Feature Request

### Description

Prior to https://github.com/prysmaticlabs/prysm/pull/7656, we had a lock around getAttPreState. The original intention was to prevent the following scenario:

- Goroutine 0: getAttPreState(checkpoint0), the state at checkpoint 0 is not cached, so we start regenerating it
- Goroutine 1: getAttPreState(checkpoint0), the state at checkpoint 0 is not cached, so we start regenerating it
- Goroutine 0: regenerated state for checkpoint is now cached
- Goroutine 1: regenerated state for checkpoint is now cached

In this scenario we repeated the work while it was already in progress.

A problem is that if another call came in for a different checkpoint (goroutine 2), it would also have been blocked by the lock held by goroutine 0, even though it would not have benefited from the work completed by goroutine 0. For this reason, we removed the lock in PR https://github.com/prysmaticlabs/prysm/pull/7656.

### Describe the solution you'd like

Ideally, goroutine 1 should have waited for goroutine 0 to complete. Instead of a generic lock, we can use a multilock where the key is the epoch + root of the checkpoint. Or a hash of the checkpoint itself. This way goroutine 1 would have rightfully been blocked by goroutine 0 but goroutine 2 would not have been blocked by goroutine 0 since it is requesting a state from a different checkpoint.

### Describe alternatives you've considered

I looked into https://github.com/atedja/go-multilock, but it requires periodic cleanup of keys and is not well suited for temporary keys. I'm thinking we'll need to create our own multilock that is self cleaning.

### PR DESCRIPTION
**Pull request by rauljordan**
**What type of PR is this?**

> Feature

**What does this PR do? Why is it needed?**

This PR includes a custom multilock implementation based on [atedja/go-multilock](https://github.com/atedja/go-multilock/blob/master/multilock_test.go), which uses custom clearing logic upon mutex unlock if a lock is not being held by anyone else. This will allow us to behave smarter in `getAttPrestate` by being able to scope concurrency by checkpoint roots without excessive mutex locking, which we have evidence is blocking runtime significantly.

**Which issues(s) does this PR fix?**

Fixes #7658 and reverts #7690

### CODE REVIEW

File: beacon-chain/blockchain/process_attestation_helpers.go [hunk 1]
```diff
@@ -13,15 +13,22 @@ import (
...
 	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/mputil"
 	"github.com/prysmaticlabs/prysm/shared/params"
 )
 
 // getAttPreState retrieves the att pre state by either from the cache or the DB.
 func (s *Service) getAttPreState(ctx context.Context, c *ethpb.Checkpoint) (*stateTrie.BeaconState, error) {
+	// Use a multilock to allow scoped holding of a mutex by a checkpoint root + epoch
+	// allowing us to behave smarter in terms of how this function is used concurrently.
+	lock := mputil.NewMultilock(string(c.Root) + fmt.Sprintf("%d", c.Epoch))
```
Comment (farazdagi) [inline beacon-chain/blockchain/process_attestation_helpers.go:24]:
```suggestion
	lock := mputil.NewMultilock(fmt.Sprintf("%s%d", c.Root, c.Epoch))
```
Any reason to use concatenation?

File: beacon-chain/blockchain/process_attestation_helpers.go [hunk 2]
```diff
@@ -13,15 +13,22 @@ import (
...
 	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/mputil"
 	"github.com/prysmaticlabs/prysm/shared/params"
 )
 
 // getAttPreState retrieves the att pre state by either from the cache or the DB.
 func (s *Service) getAttPreState(ctx context.Context, c *ethpb.Checkpoint) (*stateTrie.BeaconState, error) {
+	// Use a multilock to allow scoped holding of a mutex by a checkpoint root + epoch
+	// allowing us to behave smarter in terms of how this function is used concurrently.
+	lock := mputil.NewMultilock(fmt.Sprintf("%s%d", c.Root, c.Epoch))
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/process_attestation_helpers.go:24]: Is there a faster way to key this lock? 
I wonder about the performance of this Sprintf

File: shared/mputil/multilock_test.go [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+import (
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSetup(t *testing.T) {
```
Comment (prestonvanloon) [inline shared/mputil/multilock_test.go:24]: What's this? can we remove it?

**Comment by rauljordan** [comment]
PTAL @prestonvanloon

**Review by prestonvanloon — APPROVED** [review]

**Pull request merged**